tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/connect/ValidateAMQPErrorsTest.java

```diff
@@ -0,0 +1,643 @@
+import org.apache.activemq.artemis.core.config.amqpBrokerConnectivity.AMQPBrokerConnectionAddressType;
+import org.apache.activemq.artemis.core.config.amqpBrokerConnectivity.AMQPBrokerConnectionElement;
+import org.apache.activemq.artemis.core.config.amqpBrokerConnectivity.AMQPMirrorBrokerConnectionElement;
+import org.apache.activemq.artemis.core.server.ActiveMQServer;
+import org.apache.activemq.artemis.logs.AssertionLoggerHandler;
+import org.apache.activemq.artemis.protocol.amqp.broker.ActiveMQProtonRemotingConnection;
+import org.apache.activemq.artemis.protocol.amqp.broker.ProtonProtocolManager;
+import org.apache.activemq.artemis.protocol.amqp.connect.AMQPBrokerConnection;
+import org.apache.activemq.artemis.protocol.amqp.connect.mirror.AMQPMirrorControllerSource;
+import org.apache.activemq.artemis.tests.integration.amqp.AmqpClientTestSupport;
+import org.apache.activemq.artemis.tests.util.CFUtil;
+import org.apache.activemq.artemis.tests.util.Wait;
+import org.apache.activemq.artemis.utils.SpawnedVMSupport;
+import org.apache.activemq.artemis.utils.collections.ConcurrentHashSet;
+import org.apache.activemq.transport.amqp.client.AmqpClient;
+import org.apache.activemq.transport.amqp.client.AmqpConnection;
+import org.apache.activemq.transport.amqp.client.AmqpSender;
+import org.apache.activemq.transport.amqp.client.AmqpSession;
+import org.apache.qpid.proton.amqp.Symbol;
+import org.apache.qpid.proton.amqp.messaging.Accepted;
+import org.apache.qpid.proton.amqp.transport.AmqpError;
+import org.apache.qpid.proton.amqp.transport.ErrorCondition;
+import org.apache.qpid.proton.amqp.transport.Target;
+import org.apache.qpid.proton.engine.Link;
+import org.apache.qpid.proton.engine.Receiver;
+import org.apache.qpid.proton.engine.impl.ConnectionImpl;
+import org.jboss.logging.Logger;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Test;
+
+import static java.util.EnumSet.of;
+import static org.apache.qpid.proton.engine.EndpointState.ACTIVE;
+
+/**
+ * This test will make sure the Broker connection will react accordingly to a few misconfigs and possible errors on the network of brokers and eventually qipd-dispatch.
+ */
+public class ValidateAMQPErrorsTest extends AmqpClientTestSupport {
+
+   protected static final int AMQP_PORT_2 = 5673;
+   private static final Logger logger = Logger.getLogger(ValidateAMQPErrorsTest.class);
+   protected Vertx vertx;
+
+   protected MockServer mockServer;
+
+   public static void main(String[] arg) {
+      if (arg[0].equals("testClientMissingCapabilities")) {
+         try {
+            AmqpClient client = new AmqpClient(new URI("tcp://localhost:" + AMQP_PORT), null, null);
+            AmqpConnection connection = client.connect();
+            AmqpSession session = connection.createSession();
+            // The createSender here will block. That's the reason why we are using a separate process
```

Comment:
It likely shouldnt, unless something is wrong with it or what the broker is doing. The test client is generally synchronous but its meant to thrown on failure, which is what should be happening here.

Nothing the client or broker is doing in that 'testClientMissingCapabilities' test seems like it should need a new process. There is likely an issue that should be addressed.

Reply:
the client hangs waiting the open that won't arrive

Reply:
@gemmellr the client code will execute the following code, and it will hang on sync:

```java
      final ClientFuture request = new ClientFuture();

      connection.getScheduler().execute(new Runnable() {

         @Override
         public void run() {
            checkClosed();
            sender.setStateInspector(getStateInspector());
            sender.open(request);
            pumpToProtonTransport(request);
         }
      });

      request.sync();
```

I thought this would be a problem for the test client only and I wasn't bothered about fiixing it on the specific case here.

Since you have a lot more experience with the client side, perhaps I should use a different client and expect different results?

Reply:
I wasnt saying it doesnt, just that I dont think it should at that point given what the test is doing and the resulting link refusal that I would expect to follow from it. It should see a link refusal and throw.

It could be a bug in the test (seems very unlikely, not much room given the simplicity), in the test clients handling of what the broker does (there are other tests doing link-refusal though), or perhaps most likely it could be a bug in the brokers actual handling of the situation. Perhaps both test client and broker, given what is happening in the end. Either way, I'm pretty sure there is one that is being pointed out by this happening, and so effectively just being covered up by making the test spawn a process.

I'll take a look later after lunch.

Reply:
@gemmellr this is not doing the link refusal you mentioned.. it's just sending an error through the connection, setting the exception and sending a close frame to the client.

Reply:
@gemmellr I added a commit to be squashed.  https://github.com/clebertsuconic/activemq-artemis/commit/47836e019df39b6642418853dbcbd4be628812f1

I tried to copy the code from one of your tests doing refusal... and it still blocked.. I then called link.close and link.setTarget(null) and it worked.

if you could take a look please?

Reply:
I didnt see your comment until after I already looked at it and figured that out hehe. The test client doesnt handle being closed while waiting on other expected bits (not too surprsing, its based on simplified guts of the jms client from years ago, and lacks the upper layer that normally handles that stuff).

Having looked at your change, I still wouldn't say it is doing quite the right thing, as it doesnt include the error detail in the detach and still kills the connection after. I've modified the code and test to make it just refuse the link and include the error detail, and make the test verify that detail is present as expected. You can see the changes in https://github.com/gemmellr/activemq-artemis/commit/60789b23502e058fc3273372e7a8c24c4d9aa8a5.

(I didnt check if the alterations affect any other tests; if id does they need fixing).

Reply:
@gemmellr thanks Robbie.. thanks Again...

I will squash the commits into the code changes one...

Reply:
I will mark this conversation as resolved (just because I'm managing what's still to be done here.. if you need anything else on this subject.. please reopen the conversation.. (or ask me to do it in case GitHub doesn't give you the option to do so).